File: Array/linchuanxin/Algorithm/day33_面试题_02_08_环路检测/day33_leetcode_interview0208.java
package day33;

public class day33_leetcode_interview0208 {
    class ListNode {
      int val;
      ListNode next;
      ListNode(int x) {
          val = x;
          next = null;
      }
    }

    public ListNode detectCycle(ListNode head) {
        ListNode fast = head;
        ListNode slow = head;

        while (fast!=null){
            fast = fast.next;
            if(fast!=null){
                fast = fast.next;
                slow = slow.next;
            }

            if(fast == slow){
                break;
            }
        }

        if(fast != slow)
            return null;

        fast = head;
        while (fast!=slow){
            fast = fast.next;
            slow = slow.next;
        }

        return slow;
    }

}

File: Array/linchuanxin/Algorithm/day33_面试题_02_08_环路检测/day33_leetcode_interview0208.c
//
// Created by LCX on 2020/7/3.
//

#include <stdio.h>
#include <malloc.h>
struct ListNode{
    int val;
    struct ListNode *next;
};

struct ListNode *detectCycle(struct ListNode *head)
{
     struct ListNode *fast = head;
     struct ListNode *slow = head;

     while (fast != NULL)
     {
         fast = fast->next;
         if(fast != NULL)
         {
             fast = fast->next;
             slow = slow->next;
         }
         if (fast == slow)
             break;
     }

     if (fast != slow)
         return NULL;
     fast = head;
     while (fast != slow)
     {
         fast =fast->next;
         slow = slow->next;
     }

    return slow;

}